### PR TITLE
support UDP and multicast connections in SITL

### PR DESCRIPTION
**Pull request by tridge**
This adds support for UDP and multicast connections. These are useful for testing swarming setups, allowing for direct vehicle<->vehicle communications in SITL

syntax:
             udpclient:127.0.0.1
             udpclient:127.0.0.1:14550
             mcast:
             mcast:239.255.145.50:14550

this goes with the following pymavlink PR:
https://github.com/ArduPilot/pymavlink/pull/249

### CODE REVIEW

File: libraries/AP_HAL_SITL/UARTDriver.cpp [hunk 1]
```diff
@@ -343,6 +360,124 @@ void UARTDriver::_tcp_start_client(const char *address, uint16_t port)
...
+    }
+    ret = fcntl(_fd, F_SETFD, FD_CLOEXEC);
+    if (ret == -1) {
+        fprintf(stderr, "fcntl failed on setting FD_CLOEXEC - %s\n", strerror(errno));
+        exit(1);
+    }
+
+    // try to setup for broadcast, this may fail if insufficient privileges
+    int one = 1;
+    setsockopt(_fd,SOL_SOCKET,SO_BROADCAST,(char *)&one,sizeof(one));
```
Comment (khancyr) [inline libraries/AP_HAL_SITL/UARTDriver.cpp]: nitpick , miss some space here

**Comment by khancyr** [comment]
Very nice ! 
But is seems to be missing SOCK_CLOEXEC to allow reboot.
Otherwise, we should have a way in the code to disable mavlink rooting on some uart as on swarm configuration, using multicast or broadcast will flood GCS and vehicle with numerous mavlink spam !

**Comment by tridge** [comment]
@khancyr thanks, I've added CLOEXEC

File: libraries/AP_HAL/utility/packetise.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    uint8_t min_length = (b == MAVLINK_STX_MAVLINK1)?8:12;
+    // this looks like a MAVLink packet - try to write on
+    // packet boundaries when possible
+    if (n < min_length) {
+        // we need to wait for more data to arrive
+        return 0;
+    }
+
+    // the length of the packet is the 2nd byte, and mavlink
+    // packets have a 6 byte header plus 2 byte checksum,
```
Comment (WickedShell) [inline libraries/AP_HAL/utility/packetise.cpp]: The comment is wrong for MAVLink2, and probably belongs up next to the declaration of `min_length`

Comment (tridge) [reply libraries/AP_HAL/utility/packetise.cpp]: fixed, thanks

**Pull request merged**